src/Microsoft.Extensions.DependencyInjection/ServiceCollection.cs

```diff
@@ -63,6 +65,11 @@ public IEnumerator<ServiceDescriptor> GetEnumerator()
 
         void ICollection<ServiceDescriptor>.Add(ServiceDescriptor item)
         {
+            var registeredServiceDescriptor = _descriptors.FirstOrDefault(s => s.ServiceType == item.ServiceType);
+            if (registeredServiceDescriptor != null)
+            {
+                throw new InvalidOperationException($"There is already descriptor with service type '{registeredServiceDescriptor.ServiceType}' registered.");
```

Comment:
Should we consider an overload that keeps the old replace behavior and just removes the old registration if its found before Adding?

`Replace(item)` or `Add(item, replace: true)`